File: ExpensivePredicatesBench.java
import java.util.HashMap;
import java.util.regex.Pattern;

public class ExpensivePredicatesBench {

    static class DispatchCache {
        HashMap<String, java.util.function.Function<String, String>>[] entries;
        int nextIdx = 0;
        long hits = 0;
        long misses = 0;

        @SuppressWarnings("unchecked")
        DispatchCache() {
            entries = new HashMap[8];
            for (int i = 0; i < 8; i++) {
                entries[i] = new HashMap<>();
            }
        }

        synchronized java.util.function.Function<String, String> lookup(String key) {
            for (int i = 0; i < 8; i++) {
                if (entries[i].containsKey(key)) {
                    hits++;
                    return entries[i].get(key);
                }
            }
            misses++;
            return null;
        }

        synchronized void insert(String key, java.util.function.Function<String, String> fn) {
            entries[nextIdx].clear();
            entries[nextIdx].put(key, fn);
            nextIdx = (nextIdx + 1) % 8;
        }

        synchronized void reset() {
            hits = 0;
            misses = 0;
            nextIdx = 0;
            for (int i = 0; i < 8; i++) {
                entries[i].clear();
            }
        }
    }

    // Expensive predicates with regex matching
    static Pattern EMAIL_PATTERN = Pattern.compile("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$");
    static Pattern JSON_PATTERN = Pattern.compile("^[\\s]*[\\{\\[].*[\\}\\]][\\s]*$");
    static Pattern IP_PATTERN = Pattern.compile("^(\\d{1,3}\\.){3}\\d{1,3}$");
    static Pattern URL_PATTERN = Pattern.compile("^https?://[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}(/[a-zA-Z0-9._~:/?#\\[\\]@!$&'()*+,;=-]*)?$");
    static Pattern ALPHANUM_PATTERN = Pattern.compile("^[a-zA-Z0-9]+$");

    static boolean isEmail(String x) {
        return EMAIL_PATTERN.matcher(x).matches();
    }

    static boolean isJSON(String x) {
        return JSON_PATTERN.matcher(x).matches();
    }

    static boolean isIP(String x) {
        return IP_PATTERN.matcher(x).matches();
    }

    static boolean isURL(String x) {
        return URL_PATTERN.matcher(x).matches();
    }

    static boolean isAlphaNum(String x) {
        return ALPHANUM_PATTERN.matcher(x).matches();
    }

    static String clauseEmail(String x) { return "email"; }
    static String clauseJSON(String x) { return "json"; }
    static String clauseIP(String x) { return "ipaddress"; }
    static String clauseURL(String x) { return "url"; }
    static String clauseAlphaNum(String x) { return "alphanumeric"; }
    static String clauseUnknown(String x) { return "unknown"; }

    // Uncached dispatch
    static String dispatchUncached(String x) {
        if (isEmail(x)) return clauseEmail(x);
        if (isJSON(x)) return clauseJSON(x);
        if (isIP(x)) return clauseIP(x);
        if (isURL(x)) return clauseURL(x);
        if (isAlphaNum(x)) return clauseAlphaNum(x);
        return clauseUnknown(x);
    }

    // Cached dispatch
    static String dispatchCached(String x, DispatchCache cache) {
        String key = x.length() > 20 ? x.substring(0, 20) : x;
        java.util.function.Function<String, String> cached = cache.lookup(key);

        if (cached != null) {
            return cached.apply(x);
        }

        java.util.function.Function<String, String> fn;
        if (isEmail(x)) fn = ExpensivePredicatesBench::clauseEmail;
        else if (isJSON(x)) fn = ExpensivePredicatesBench::clauseJSON;
        else if (isIP(x)) fn = ExpensivePredicatesBench::clauseIP;
        else if (isURL(x)) fn = ExpensivePredicatesBench::clauseURL;
        else if (isAlphaNum(x)) fn = ExpensivePredicatesBench::clauseAlphaNum;
        else fn = ExpensivePredicatesBench::clauseUnknown;

        cache.insert(key, fn);
        return fn.apply(x);
    }

    public static void main(String[] args) throws InterruptedException {
        int numCalls = 100000;

        System.out.println("================================");
        System.out.println("OpenJDK C2 Expensive Predicates Dispatch Caching Benchmark");
        System.out.println("================================");
        System.out.println("Test data: " + numCalls + " calls with expensive predicates");
        System.out.println("  Predicates: email, JSON, IP address, URL, alphanumeric");
        System.out.println("  Predicate cost: 1-3 microseconds each (regex matching)");
        System.out.println("  Uncached dispatch cost: 5-15 microseconds");
        System.out.println("  Cache benefit potential: Massive if expensive predicates save checks\n");

        String[] testData = new String[numCalls];
        String[] patterns = {
            "user@example.com",
            "{\"key\": \"value\"}",
            "192.168.1.1",
            "https://example.com/path?query=value",
            "abc123def456"
        };

        for (int i = 0; i < numCalls; i++) {
            testData[i] = patterns[i % 5];
        }

        // Warmup
        System.out.println("Warming up JIT (10,000 calls)...");
        for (int i = 0; i < 10000; i++) {
            dispatchUncached(testData[i % numCalls]);
        }

        DispatchCache cache = new DispatchCache();
        for (int i = 0; i < 10000; i++) {
            dispatchCached(testData[i % numCalls], cache);
        }
        System.out.println("Warmup complete.\n");

        // Uncached benchmark
        System.out.println("=== Uncached Expensive Dispatch (3 iterations) ===");
        long[] uncachedTimes = new long[3];
        for (int run = 0; run < 3; run++) {
            long start = System.nanoTime();
            for (int i = 0; i < numCalls; i++) {
                dispatchUncached(testData[i]);
            }
            long elapsed = System.nanoTime() - start;
            uncachedTimes[run] = elapsed;
            double seconds = elapsed / 1e9;
            double usPerCall = (elapsed / 1e3) / numCalls;
            System.out.printf("  Run %d: %.4f seconds (%.2f us/call)%n", run + 1, seconds, usPerCall);
        }

        // Cached benchmark
        System.out.println("\n=== Cached Expensive Dispatch (3 iterations) ===");
        long[] cachedTimes = new long[3];
        for (int run = 0; run < 3; run++) {
            cache.reset();
            long start = System.nanoTime();
            for (int i = 0; i < numCalls; i++) {
                dispatchCached(testData[i], cache);
            }
            long elapsed = System.nanoTime() - start;
            cachedTimes[run] = elapsed;
            double seconds = elapsed / 1e9;
            double usPerCall = (elapsed / 1e3) / numCalls;
            System.out.printf("  Run %d: %.4f seconds (%.2f us/call)%n", run + 1, seconds, usPerCall);
        }

        // Stats
        System.out.println("\nCached Dispatch Stats:");
        System.out.println("  Cache hits: " + cache.hits);
        System.out.println("  Cache misses: " + cache.misses);
        long total = cache.hits + cache.misses;
        if (total > 0) {
            double hitRate = 100.0 * cache.hits / total;
            System.out.printf("  Hit rate: %.4f%%%n", hitRate);
        }

        // Summary
        long avgUncached = (uncachedTimes[0] + uncachedTimes[1] + uncachedTimes[2]) / 3;
        long avgCached = (cachedTimes[0] + cachedTimes[1] + cachedTimes[2]) / 3;

        double usUncached = (avgUncached / 1e3) / numCalls;
        double usCached = (avgCached / 1e3) / numCalls;
        double ratio = (double) avgCached / avgUncached;

        System.out.println("\n================================");
        System.out.printf("Average uncached: %.2f us/call%n", usUncached);
        System.out.printf("Average cached: %.2f us/call%n", usCached);
        System.out.printf("Speedup/Slowdown ratio: %.2f%%%n", ratio * 100);

        if (ratio < 1.0) {
            System.out.printf("SPEEDUP: %d%% faster%n", (int) ((1.0 - ratio) * 100));
        } else {
            System.out.printf("SLOWDOWN: %d%% slower%n", (int) ((ratio - 1.0) * 100));
        }

        System.out.println("\nBreak-even analysis:");
        System.out.println("  Predicate cost: 5 to 15 microseconds (expensive)");
        System.out.println("  Cache overhead: Varies with JIT optimization");
        System.out.println("  Theoretical: Should see speedup if expensive predicates save checks");
        System.out.println("================================");
    }
}

File: docs/caching/expensive_predicates_bench.py
#!/usr/bin/env python3
"""
Expensive Predicates Dispatch Caching Benchmark for CPython and PyPy
Tests dispatch with expensive regex pattern matching predicates
"""

import re
import time
from typing import Dict, Callable, Optional

class DispatchCache:
    """Round-robin 8-slot LRU cache for dispatch decisions"""

    def __init__(self):
        self.entries = [dict() for _ in range(8)]
        self.next_idx = 0
        self.hits = 0
        self.misses = 0

    def lookup(self, key: str) -> Optional[Callable]:
        for entry in self.entries:
            if key in entry:
                self.hits += 1
                return entry[key]
        self.misses += 1
        return None

    def insert(self, key: str, fn: Callable) -> None:
        self.entries[self.next_idx].clear()
        self.entries[self.next_idx][key] = fn
        self.next_idx = (self.next_idx + 1) % 8

    def reset(self) -> None:
        self.hits = 0
        self.misses = 0
        self.next_idx = 0
        for entry in self.entries:
            entry.clear()

# Expensive predicates using regex (1-3 microseconds each)
EMAIL_REGEX = re.compile(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$')
JSON_REGEX = re.compile(r'^[\s]*[\{\[].*[\}\]][\s]*$')
IP_REGEX = re.compile(r'^(\d{1,3}\.){3}\d{1,3}$')
URL_REGEX = re.compile(r'^https?://[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}(/[a-zA-Z0-9._~:/?#\[\]@!$&\'()*+,;=-]*)?$')
ALPHANUM_REGEX = re.compile(r'^[a-zA-Z0-9]+$')

def predicate_is_email(x: str) -> bool:
    """Check if x is a valid email (expensive regex, 2-3 microseconds)"""
    return bool(EMAIL_REGEX.match(x))

def predicate_is_json(x: str) -> bool:
    """Check if x looks like JSON (expensive regex, 1-2 microseconds)"""
    return bool(JSON_REGEX.match(x))

def predicate_is_ip(x: str) -> bool:
    """Check if x is an IP address (expensive regex, 2-3 microseconds)"""
    return bool(IP_REGEX.match(x))

def predicate_is_url(x: str) -> bool:
    """Check if x is a URL (expensive regex, 2-3 microseconds)"""
    return bool(URL_REGEX.match(x))

def predicate_is_alphanum(x: str) -> bool:
    """Check if x is alphanumeric only (expensive regex, 1-2 microseconds)"""
    return bool(ALPHANUM_REGEX.match(x))

# Clause functions
def clause_email(x: str) -> str:
    return "email"

def clause_json(x: str) -> str:
    return "json"

def clause_ip(x: str) -> str:
    return "ipaddress"

def clause_url(x: str) -> str:
    return "url"

def clause_alphanum(x: str) -> str:
    return "alphanumeric"

def clause_unknown(x: str) -> str:
    return "unknown"

# Uncached dispatch: test all predicates (5-15 microseconds per call)
def dispatch_uncached(x: str) -> str:
    if predicate_is_email(x):
        return clause_email(x)
    elif predicate_is_json(x):
        return clause_json(x)
    elif predicate_is_ip(x):
        return clause_ip(x)
    elif predicate_is_url(x):
        return clause_url(x)
    elif predicate_is_alphanum(x):
        return clause_alphanum(x)
    else:
        return clause_unknown(x)

# Cached dispatch
def dispatch_cached(x: str, cache: DispatchCache) -> str:
    key = x[:20] if len(x) > 20 else x
    cached = cache.lookup(key)

    if cached is not None:
        return cached(x)

    # Find matching predicate
    if predicate_is_email(x):
        fn = clause_email
    elif predicate_is_json(x):
        fn = clause_json
    elif predicate_is_ip(x):
        fn = clause_ip
    elif predicate_is_url(x):
        fn = clause_url
    elif predicate_is_alphanum(x):
        fn = clause_alphanum
    else:
        fn = clause_unknown

    cache.insert(key, fn)
    return fn(x)

def run_benchmark(implementation_name: str = "CPython") -> None:
    num_calls = 100000

    print("================================")
    print(f"{implementation_name} Expensive Predicates Dispatch Caching Benchmark")
    print("================================")
    print(f"Test data: {num_calls} calls with expensive predicates")
    print("  Predicates: email, JSON, IP address, URL, alphanumeric")
    print("  Predicate cost: 1-3 microseconds each (regex matching)")
    print("  Uncached dispatch cost: 5-15 microseconds")
    print("  Cache benefit potential: Massive if expensive predicates save checks\n")

    patterns = [
        "user@example.com",
        '{"key": "value"}',
        "192.168.1.1",
        "https://example.com/path?query=value",
        "abc123def456",
    ]

    # Create test data
    test_data = [patterns[i % 5] for i in range(num_calls)]

    # Warmup
    print("Warming up (10,000 calls)...")
    for i in range(10000):
        dispatch_uncached(test_data[i % num_calls])

    cache = DispatchCache()
    for i in range(10000):
        dispatch_cached(test_data[i % num_calls], cache)
    print("Warmup complete.\n")

    # Uncached benchmark
    print("=== Uncached Expensive Dispatch (3 iterations) ===")
    uncached_times = []
    for run in range(3):
        start = time.perf_counter()
        for item in test_data:
            dispatch_uncached(item)
        elapsed = time.perf_counter() - start
        uncached_times.append(elapsed)
        us_per_call = (elapsed * 1e6) / num_calls
        print(f"  Run {run + 1}: {elapsed:.4f} seconds ({us_per_call:.2f} us/call)")

    # Cached benchmark
    print("\n=== Cached Expensive Dispatch (3 iterations) ===")
    cached_times = []
    for run in range(3):
        cache.reset()
        start = time.perf_counter()
        for item in test_data:
            dispatch_cached(item, cache)
        elapsed = time.perf_counter() - start
        cached_times.append(elapsed)
        us_per_call = (elapsed * 1e6) / num_calls
        print(f"  Run {run + 1}: {elapsed:.4f} seconds ({us_per_call:.2f} us/call)")

    # Cache stats
    print("\nCached Dispatch Stats:")
    print(f"  Cache hits: {cache.hits}")
    print(f"  Cache misses: {cache.misses}")
    total = cache.hits + cache.misses
    if total > 0:
        hit_rate = 100.0 * cache.hits / total
        print(f"  Hit rate: {hit_rate:.4f}%")

    # Summary
    avg_uncached = sum(uncached_times) / 3
    avg_cached = sum(cached_times) / 3

    us_uncached = (avg_uncached * 1e6) / num_calls
    us_cached = (avg_cached * 1e6) / num_calls
    ratio = avg_cached / avg_uncached

    print("\n================================")
    print(f"Average uncached: {us_uncached:.2f} us/call")
    print(f"Average cached: {us_cached:.2f} us/call")

    if ratio < 1.0:
        print(f"SPEEDUP: {int((1.0 - ratio) * 100)}% faster with caching")
    else:
        print(f"SLOWDOWN: {int((ratio - 1.0) * 100)}% slower with caching")

    print("\nBreak-even analysis:")
    print("  Predicate cost: 5 to 15 microseconds (expensive)")
    print("  Cache overhead: negligible at this scale")
    print("  Theoretical: Should see speedup if expensive predicates save checks")
    print("================================\n")

if __name__ == "__main__":
    import sys
    impl_name = sys.argv[1] if len(sys.argv) > 1 else "CPython"
    run_benchmark(impl_name)
